workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
+                {
+                    if (!componentIdToWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToWriterSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                numUnsatisfiedReadersOrWriters[behaviour] = readerIds.Count + writerIds.Count;
+            }
         }
 
-        public List<ReaderWriterImpl> GetReadersWriters(uint componentId)
+
+        public HashSet<IReaderInternal> GetReadersWriters(uint componentId)
         {
-            return null;
+            return componentIdToCreatedReadersWriters[componentId];
         }
 
         public void EnableSpatialOSBehaviours()
         {
-            // inject + enable all spatialOSBehavioursToEnable
+            foreach (var behaviour in spatialOSBehavioursToEnable)
+            {
+                var dict = behaviourLibrary.InjectAllReadersWriters(behaviour, entity);
+                behaviourToCreatedReadersWriters[behaviour] = dict;
+                foreach (var idToReaderWriter in dict)
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    if (!componentIdToCreatedReadersWriters.ContainsKey(id))
+                    {
+                        componentIdToCreatedReadersWriters[id] = new HashSet<IReaderInternal>();
+                    }
+
+                    componentIdToCreatedReadersWriters[id].Add(reader);
+                }
+
+                behaviour.enabled = true;
+            }
+
+            spatialOSBehavioursToEnable.Clear();
         }
 
         public void DisableSpatialOSBehaviours()
         {
-            // disable + deinject all spatialOSBehavioursToDisable
+            foreach (var behaviour in spatialOSBehavioursToDisable)
+            {
+                behaviour.enabled = false;
+                behaviourLibrary.DeInjectAllReadersWriters(behaviour);
+                foreach (var idToReaderWriter in behaviourToCreatedReadersWriters[behaviour])
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    componentIdToCreatedReadersWriters[id].Remove(reader);
+                }
+
+                behaviourToCreatedReadersWriters.Remove(behaviour);
+            }
+
+            spatialOSBehavioursToDisable.Clear();
         }
 
         public void AddComponent(uint componentId)
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = componentIdToReaderSpatialOSBehaviours[componentId];
+            foreach (var spatialOSBehaviour in relevantReaderSpatialOSBehaviours)
+            {
+                MarkComponentReady(spatialOSBehaviour);
```

Comment:
Why not refactor `MarkComponentReady` and `MarkComponentNotReady` to accept an `IEnumerable<>` so the loops happen in the methods rather than at the call site?